tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/TensorScope.java

```diff
@@ -0,0 +1,211 @@
+/*
+  Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ==============================================================================
+ */
+package org.tensorflow;
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+
+/**
+ * A scope that can be used to manage tensor resources.  If auto-attach is used, any tensors created between a scope's
+ * creation and calling {@code close()}, that haven't been detached, are guaranteed to be closed with the scope (even if
+ * they are created in a sub-scope).  Tensors may be manually closed earlier without issue.
+ * <p>
+ * When auto-attach is true, tensors are automatically tracked on creation.  A tensor can me manually added to a scope
+ * with {@link TensorScope#attach(Tensor)} or {@link Tensor#attachToCurrentScope()}.  The tensor will then be closed when the first of it's managing scopes closes.
```

Comment:
Why would we want Tensors to be closed by the first scope closure rather than the last? If it's attached to multiple scopes then I'd expect it to still be alive if any of the scopes were alive.

Reply:
Ah, is it because of the stacking of scopes? Even though they could be stored in fields and thus mess up the logical stacking?

Reply:
Right, this situation shouldn't happen anyway. It's not clear to me that this was designed to used in fields, so let's assume we only have references on the stack. If we assume that's true, then any "inner" scopes are supposed to be closed before "outer" scopes. If this doesn't happen, there's probably some logic gone pretty wrong somewhere that this should be the least of our worries...

Reply:
Yeah, exactly, it's not intended to be used in fields, unless it's as a way to ensure that any created tensors are closed when the creator is closed, like in EagerSession.  I'm thinking about removing that anyways.